File: 6. Dijkstra Shortest Path.cpp
#include <bits/stdc++.h>
// Dijkstra's Shortest Path Algorithm
// Greedy strategy: Always expand the node with the smallest tentative distance (using a min-heap).
// Task: Given a weighted, directed graph with non-negative edge weights and a source vertex,
// find the shortest path from the source to all other vertices in the graph.

#include <functional>
#include <iostream>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

vector<long long> dijkstra(int n, vector<vector<pair<int, int>>>& adj, int s)
{
    const long long INF = 4e18;
    vector<long long> dist(n, INF);
    priority_queue<pair<long long, int>, vector<pair<long long, int>>,
                   greater<pair<long long, int>>>
        pq;
    dist[s] = 0;
    pq.push({0, s});
    while (!pq.empty())
    {
        auto [d, u] = pq.top();
        pq.pop();
        if (d != dist[u])
            continue;
        for (auto [v, w] : adj[u])
        {
            if (dist[v] > d + w)
            {
                dist[v] = d + w;
                pq.push({dist[v], v});
            }
        }
    }
    return dist;
}

int main()
{
    int n = 5, s = 0;
    vector<vector<pair<int, int>>> adj(n);
    auto add = [&](int u, int v, int w) { adj[u].push_back({v, w}); };
    add(0, 1, 10);
    add(0, 4, 5);
    add(1, 2, 1);
    add(4, 1, 3);
    add(4, 2, 9);
    add(4, 3, 2);
    add(3, 2, 4);
    add(2, 3, 6);
    add(3, 0, 7);
    auto dist = dijkstra(n, adj, s);
    for (int i = 0; i < n; ++i) cout << dist[i] << (i + 1 == n ? '\n' : ' ');
    return 0;
}

/*
Time Complexity: O((V + E) log V) — binary heap implementation.
Space Complexity: O(V + E) — adjacency list and distance array.
*/

File: 6. Dijkstra Shortest Path.py
"""
Dijkstra's Shortest Path Algorithm
Task: Given a weighted, directed graph with non-negative edge weights and a source vertex, find the
shortest path from the source to all other vertices in the graph.
Greedy strategy: Always expand the node with the smallest tentative distance (using a min-heap).
"""
import heapq

def dijkstra(n, adj, src):
    INF = 10**18
    dist = [INF]*n
    dist[src] = 0
    pq = [(0, src)]
    while pq:
        d, u = heapq.heappop(pq)
        if d != dist[u]:
            continue
        for v, w in adj[u]:
            if dist[v] > d + w:
                dist[v] = d + w
                heapq.heappush(pq, (dist[v], v))
    return dist

if __name__ == "__main__":
    n, src = 5, 0
    adj = [[] for _ in range(n)]
    def add(u, v, w):
        adj[u].append((v,w))
    add(0,1,10); add(0,4,5); add(1,2,1); add(4,1,3); add(4,2,9); add(4,3,2);
    add(3,2,4); add(2,3,6); add(3,0,7)
    print(*dijkstra(n, adj, src))
    # Time Complexity: O((V + E) log V) — binary heap.
    # Space Complexity: O(V + E).
